datafusion/expr/src/expr_schema.rs

```diff
@@ -112,7 +112,23 @@ impl ExprSchemable for Expr {
             Expr::OuterReferenceColumn(ty, _) => Ok(ty.clone()),
             Expr::ScalarVariable(ty, _) => Ok(ty.clone()),
             Expr::Literal(l) => Ok(l.data_type()),
-            Expr::Case(case) => case.when_then_expr[0].1.get_type(schema),
+            Expr::Case(case) => {
+                let then_type = case.when_then_expr[0].1.get_type(schema)?;
+                if !then_type.is_null() {
+                    return Ok(then_type);
+                }
+
+                let else_type = case
+                    .else_expr
+                    .as_ref()
+                    .map_or(Ok(DataType::Null), |e| e.get_type(schema))?;
+
+                match (then_type.clone(), else_type.clone()) {
```

Comment:
```suggestion
                match (&then_type, &else_type) {
```
I think we can avoid these clones.